drivers/regulator/regulator_gpio.c

```diff
@@ -0,0 +1,187 @@
+	struct regulator_common_data common;
+	int32_t current_volt_uv;
+};
+
+static unsigned int regulator_gpio_count_voltages(const struct device *dev)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+
+	return cfg->states_cnt / 2;
+}
+
+static int regulator_gpio_list_voltage(const struct device *dev, unsigned int idx,
+				       int32_t *volt_uv)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+
+	if (idx >= cfg->states_cnt / 2) {
+		LOG_ERR("%s: can't get list voltage for idx %u", dev->name, idx);
+		return -EINVAL;
+	}
+
+	*volt_uv = cfg->states[idx * 2];
+	return 0;
+}
+
+static int regulator_gpio_get_state_voltage(const struct device *dev, int state,
+				       int32_t *volt_uv)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+	unsigned int state_idx = 0;
+
+	for (; state_idx < cfg->states_cnt / 2; state_idx++) {
+		if (cfg->states[state_idx * 2 + 1] == state) {
+			*volt_uv = cfg->states[state_idx * 2];
+			return 0;
+		}
+	}
+
+	LOG_ERR("%s: can't get voltage for state %u", dev->name, state);
+	return -EINVAL;
+}
+
+static int regulator_gpio_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+	struct regulator_gpio_data *data = dev->data;
+	unsigned int state = 0;
+	int32_t mid = (min_uv + max_uv) / 2;
+	int32_t min_diff = INT32_MAX;
+	unsigned int best_state;
+	int ret = 0;
+
+	for (; state < cfg->states_cnt / 2; state++) {
+		int32_t diff;
+
+		if (!IN_RANGE(cfg->states[state * 2], min_uv, max_uv)) {
+			continue;
+		}
+
+		diff = mid - cfg->states[state * 2];
+		if (diff < 0) {
+			diff = -diff;
+		}
+
+		if (diff < min_diff) {
+			min_diff = diff;
+			best_state = state;
+		}
+	}
+
+	if (min_diff == INT32_MAX) {
+		LOG_ERR("%s: can't find voltage is states", dev->name);
+		return -EINVAL;
+	}
+
+	ret = gpio_pin_get_dt(&cfg->switch_volt);
+	if (ret < 0) {
+		LOG_ERR("%s: can't get pin state", dev->name);
+		return ret;
+	}
+
+	if (ret != cfg->states[best_state * 2 + 1]) {
+		ret = gpio_pin_set_dt(&cfg->switch_volt, cfg->states[best_state * 2 + 1]);
+		if (ret < 0) {
+			LOG_ERR("%s: can't set pin state", dev->name);
+			return ret;
+		}
+	}
+
+	data->current_volt_uv = cfg->states[best_state * 2];
+	return 0;
+}
+
+static int regulator_gpio_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_gpio_data *data = dev->data;
+
+	*volt_uv = data->current_volt_uv;
+	return 0;
+}
+
+static const struct regulator_driver_api regulator_gpio_api = {
+	.set_voltage = regulator_gpio_set_voltage,
+	.get_voltage = regulator_gpio_get_voltage,
+	.count_voltages = regulator_gpio_count_voltages,
+	.list_voltage = regulator_gpio_list_voltage,
+};
+
+static int regulator_gpio_init(const struct device *dev)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+	struct regulator_gpio_data *data = dev->data;
+	int ret;
+
+	regulator_common_data_init(dev);
+
+	if (!device_is_ready(cfg->switch_volt.port)) {
```

Comment:
`gpio_is_ready_dt`

Reply:
this piece of code as the same as in other regulators with gpios